Change in logic of `from_sub_trees_as_trees` constructor

storojs72 opened the pull request:
This PR is rebased on top of https://github.com/filecoin-project/merkletree/pull/100 (that introduced framework for integration testing). It refactors `from_sub_trees_as_trees` constructor of MerkleTree (top-tree) instance in a more generic way, e.g. values of `SubTreeArity` and `TopTreeArity` can be either even or odd (previously only even numbers were supported and, with odd numbers, mentioned constructor panicked). This change has been evaluated using high-level seal-lifecycle tests for production sectors ([32gib](https://github.com/filecoin-project/rust-fil-proofs/blob/master/filecoin-proofs/tests/api.rs#L232) and [64gib](https://github.com/filecoin-project/rust-fil-proofs/blob/master/filecoin-proofs/tests/api.rs#L264)). Logs of successful executing mentioned tests on proofs-buildbot are included.

vmx commented on src/merkle.rs:
```diff
@@ -1810,6 +1810,37 @@ impl Element for [u8; 32] {
+    let sub_tree_arity = SubTreeArity::to_usize();
+    let base_tree_arity = BaseTreeArity::to_usize();
+
+    let base_tree_len = match get_merkle_tree_len(leaves, base_tree_arity) {
```
I think this can be written as `let base_tree_len = get_merkle_tree_len(leaves, base_tree_arity)?;`

storojs72 replied:
Definitely! Thanks!

vmx commented on src/merkle.rs:
```diff
@@ -1810,6 +1810,37 @@ impl Element for [u8; 32] {
     }
 }
 
+pub fn get_merkle_tree_len_generic<
```
Some documentation would be good as it is quite interesting in which cases which values are returned.

storojs72 replied:
Added missing explanation

vmx commented on src/merkle.rs:
```diff
@@ -2041,6 +2074,13 @@ fn test_get_merkle_tree_methods() {
     assert!(get_merkle_tree_len(1, 2).is_err());
     assert!(get_merkle_tree_len(4, 16).is_err());
     assert!(get_merkle_tree_len(1024, 11).is_err());
+    assert!(get_merkle_tree_len_generic::<U4, U0, U0>(16).is_ok());
+    assert!(get_merkle_tree_len_generic::<U1, U0, U0>(3).is_ok());
```
You could also use `assert_eq!()` and test for the actual return value.

storojs72 replied:
Sure. Added mentioned tests. They will never be redundant

vmx commented on src/test_merkle_constructors.rs:
```diff
@@ -0,0 +1,338 @@
+
+/// Dataset generators. It is assumed that every generator will produce dataset with particular length, equal to leaves parameter
+fn generate_vector_of_elements<E: Element>(leaves: usize) -> Vec<E> {
+    let mut dataset = Vec::<E>::new();
+    for index in 0..leaves {
+        // we are ok with usize -> u8 conversion problems, since we need just predictable dataset
+        let vector: Vec<u8> = (0..E::byte_len()).map(|x| (index + x) as u8).collect();
+        let element = E::from_slice(vector.as_slice());
+        dataset.push(element);
+    }
+    dataset
```
I'd also write the outer loop with an functional-style iterator. This way you don't need mutability and it's easy to spot that you operate on all elements. I would write this function as:

```rust
    (0..leaves).map(|index| {
        // we are ok with usize -> u8 conversion problems, since we need just predictable dataset
        let vector: Vec<u8> = (0..E::byte_len()).map(|x| (index + x) as u8).collect();
        E::from_slice(vector.as_slice())
    })
```

storojs72 replied:
Added this refactoring. Thanks

vmx commented on src/test_merkle_constructors.rs:
```diff
@@ -0,0 +1,338 @@
+    dataset
+}
+fn generate_vector_of_usizes(leaves: usize) -> Vec<usize> {
+    let mut dataset = Vec::with_capacity(leaves);
```
Same here, I'd use `map()`.

storojs72 replied:
ditto. Thanks

vmx reviewed (commented):
Only minor things.

cryptonemo commented on src/merkle.rs:
```diff
@@ -447,7 +447,7 @@ impl<
 
         let store = S::new_from_slice(tree_len, &data).context("failed to create data store")?;
-        let root = store.read_at(data.len() - 1)?;
+        let root = store.read_at(store.len() - 1)?;
```
Are you sure these are compatible?  I thought one was in element counts and one was in data/byte counts(?)

storojs72 replied:
Thanks, for pointing on this. This is a place, where I'm not sure completely, TBH. But two following thoughts caused me to introduce this change:
- I have double-checked that `from_tree_slice` constructor doesn't have tests neither in this repository, nor in `rust_fil_proofs` - so it is impossible to check compatibility actually;
- `data` from this constructor is actually a byte slice that, as far as I understand this, represents serialised storage in some custom format. But `store` is an instance of storage, created after deserialising `data` via `S::new_from_slice`, and if we want to get last element from store (which is root of tree), we need to use indexing of store - not data. 

I added tests for this constructor, where base binary tree with vector storage is instantiated with 4 leaves (each leaf is a 16-bytes array) and basic functionality of the tree is evaluated. This tree has 7 elements in total (4 leaves + 2 nodes + 1 root), which gives 7 * 16 = 112 bytes slice for serialised tree (our `data`). With previous version of code I was getting panic:

```
thread 'test_merkle_constructors::test_from_tree_slice_group' panicked at 'index out of bounds: the len is 7 but the index is 111', src/store/vec.rs:86:12
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_merkle_constructors::test_from_tree_slice_group ... FAILED
```
specifically at this place, when root was being read. And it is reasonable, since we use `data` which is 112 bytes slice, instead of `store` (of `VecStore` type) which is actually vector of elements (`Element` type, which in my case is `Item` - 16-bytes array), so according to the logic in mind, we need to take last `Item` from our `store`, which has index: `[store.len() - 1]`.

storojs72 commented:
@vmx, thanks for initial review. I will introduce your remarks, once finishing with other constructors

cryptonemo reviewed (approved):
I like the map usage suggestions, but looks good overall

storojs72 commented:
@cryptonemo , take a look please at https://github.com/filecoin-project/merkletree/pull/98/commits/46fb4974767c7a95a8f448d5ef0808af5a9d98a9. Here I introduced changes into `from_sub_tree_as_trees` construction of compound-compound tree. Previously it didn't have tests, so I couldn't definitely understand its intention. My suspicion regarding this is that given an ordered set of base trees, this constructor groups base trees according to provided sub-tree arity, then instantiates at first set of compound trees and finally combines them into compound-compound tree. I don't know for sure whether `split_off` was (and is now) used properly, so now I changed logic a bit, according to my understanding and added proper test for it. Let me know please, if I missed something

storojs72 commented:
I noted that currently we have +1000 lines of code approximately. After introducing yet missing tests (that will add some more lines, but not too much, I think) we will have integrated test suite with more or less full coverage of constructors and time will come for cleaning. Considering that we have more than 1600 lines in test_xor128.rs and as it seems to me it has less coverage, that is an acceptable result - because most of the tests from test_xor128.rs will be removed while cleaning

storojs72 commented on src/test_merkle_constructors.rs:
```diff
@@ -0,0 +1,338 @@
+use crate::store::{Store, VecStore};
+use crate::test_common::{Item, Sha256Hasher, XOR128};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use typenum::{Unsigned, U0, U2, U3, U5};
```
Got it. Introduced this change. Thanks for pointing. Will try to keep in mind

storojs72 commented:
@vmx , @cryptonemo , I've finally come up with all constructors covered. And I added proposed changes. Take please a look once again. In a meanwhile, I will try to experiment with tests in `rust-fil-proofs`, in order to evaluate `split_off`-based changes in `from_sub_tree_as_trees` constructor, which is questionable. Thanks!

vmx reviewed (approved):
I haven't reviewed all constructors closely. If I should, please let me know.

I saw a few `for` loops that could be `map()`s, but in some cases I can also imagine that the type annotations might become crazy. So I'm OK with just keeping the current state.

storojs72 commented:
> I can also imagine that the type annotations might become crazy

That's true, you are right! I also tried to replace for loops with functional constructions where possible

storojs72 commented:
@cryptonemo, I narrow down the scope of this PR to discussion of `from_sub_trees_as_trees` constructor, by introducing https://github.com/filecoin-project/merkletree/pull/100. It allows unblock addition of integration tests.

storojs72 commented:
I upload log of successful executing [test_seal_lifecycle_32gib_porep_id_v1_1_top_8_8_0_api_v1_1](https://github.com/filecoin-project/rust-fil-proofs/blob/master/filecoin-proofs/tests/api.rs#L240) test from `rust-fil-proofs`: [32gib.txt](https://github.com/filecoin-project/merkletree/files/8547654/32gib.txt). I'm still working with executing similar 64gib test in order to check compatibility of this change with sectors used in production

storojs72 commented:
Uploading full log of successful executing [test_seal_lifecycle_64gib_porep_id_v1_1_top_8_8_2_api_v1_1](https://github.com/filecoin-project/rust-fil-proofs/blob/master/filecoin-proofs/tests/api.rs#L264) test:
[64gib.txt](https://github.com/filecoin-project/merkletree/files/8640253/64gib_success.txt)

vmx reviewed (approved):
The new code is way easier to follow. I think it could even be more clear using `map` instead of for loops.

vmx commented:
I forgot to add that I think it would be good to squash it into a single commit with a good commit message explaining that the old code was wrong etc.
